# OBJECTIVE :-To See a significant difference in marks between A,B And C. In this case if the variable has a normal distribution then ANOVA TEST is used.If the variable doesn't have that then KRUSKAL WALIS TEST is used.
## The libraries in python consist of Functions and Datasets.The name of the dataset that we used here is IRIS which is present inside the seaborn library. 
### 1.Import the pandas and seaborn libraries.
### 2.To import the dataset from the seaborn library,we use 'sns.load_dataset' like we use read_csv earlier.
### 3.The function changes depending on the library from which the dataset has been called.
## The null hypothesis is there is no significant difference in the sepal length across the three species.
## Alternate hypotheis is there is a significant difference in sepal length across the three species.

In [2]:
import pandas as pd
import seaborn as sns

In [3]:
iris=sns.load_dataset('iris') #The function and dataset is inside seaborn library.
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

# To check the datatype of the variables iris.info() is used.


In [4]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


# To convert the variable from object datatype to categorical:The variable species need to be converted into categorical.

In [5]:
iris['species']=pd.Categorical(iris['species']) #Categorical is the function in the seaborn library.We can also create new variable in the dataset.

In [6]:
iris.info() #the variable species is getting converted into Category from Object.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   sepal_length  150 non-null    float64 
 1   sepal_width   150 non-null    float64 
 2   petal_length  150 non-null    float64 
 3   petal_width   150 non-null    float64 
 4   species       150 non-null    category
dtypes: category(1), float64(4)
memory usage: 5.1 KB


#### Categorical variable takes less variable as compared to object variable.It is mandatory to have categorical datatype for the ANOVA TEST to run.

# To check the categories for species:

In [7]:
iris['species'].unique() #This function tells that there are three distinct categories in the categorical variable.

['setosa', 'versicolor', 'virginica']
Categories (3, object): ['setosa', 'versicolor', 'virginica']

In [8]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
import numpy as np

# To perform ANOVA and check the hypothesis.

In [9]:
formula='sepal_width ~ C(species)' # C stands for categorical.
model=ols(formula,iris).fit() #iris here is dataframe.ols stands for Ordinary Least Square.
aov_table=anova_lm(model)
print(aov_table)



               df     sum_sq   mean_sq         F        PR(>F)
C(species)    2.0  11.344933  5.672467  49.16004  4.492017e-17
Residual    147.0  16.962000  0.115388       NaN           NaN


## The value calculated from python is more than the critical value from the table so the null hypothesis is rejected and there is a significant difference in the sepal width across the three species.

# Question 1:-Use the dataset hsb2.csv.Replace 1 with b.tech,2 with m.tech and 3 with research.The objective is to see is there a significance difference in the reading scale of the students between the three kinds of programs.

In [13]:
df1=pd.read_csv('F:\hsbw2.csv')

In [14]:
df1['prog'].unique()

array(['b.tech', 'research', 'm.tech'], dtype=object)

In [15]:
formula='read ~ C(prog)' # C stands for categorical.
model=ols(formula,df1).fit() #iris here is dataframe.ols stands for Ordinary Least Square.
aov_table=anova_lm(model)
print(aov_table)

             df       sum_sq      mean_sq          F        PR(>F)
C(prog)     2.0   3716.86127  1858.430635  21.282348  4.283274e-09
Residual  197.0  17202.55873    87.322633        NaN           NaN


# The null hypothesis is rejected as the value of f from the table is 3.04 and the calculated value is 21.28 which is greater than the value from the table.
# There are four categories from the variable race north south east west.check the hypothesis is there a significant difference in the math with the race difference.

In [18]:
df2=pd.read_csv('F:\hsby2.csv')

In [20]:
df2['race'].unique()

array(['west', 'east', 'north', 'south'], dtype=object)

In [21]:
formula='math ~ C(race)' # C stands for categorical.
model=ols(formula,df2).fit() #iris here is dataframe.ols stands for Ordinary Least Square.
aov_table=anova_lm(model)
print(aov_table)

             df        sum_sq     mean_sq         F    PR(>F)
C(race)     3.0   1842.140193  614.046731  7.703265  0.000068
Residual  196.0  15623.654807   79.712525       NaN       NaN


# The null hypothesis is rejected here also.

# KRUSKAL WALIS TEST:-
## If the variable doesn't have the normal distribution then used Kruskal Walis Test.

In [24]:
from scipy.stats import kruskal
stat,p=kruskal(iris['sepal_width'][iris['species']=='setosa'],
              iris['sepal_width'][iris['species']=='versicolor'],
              iris['sepal_width'][iris['species']=='virginica'])
print('stat=%.3f,p=%.3f'%(stat,p))
if p>0.05:
    print('Probably the same distribution')
else:
    print('Probably different distribution')

stat=63.571,p=0.000
Probably different distribution


# Replace 1 with low,2 with middle,3 with high for the ses.do the kruskal walis test to see is there a significant difference in the science iq with students from different socio-economic status(ses).

In [28]:
df3=pd.read_csv('F:\hsba2.csv')

In [29]:
from scipy.stats import kruskal
stat,p=kruskal(df3['science'][df3['ses']=='low'],
              df3['science'][df3['ses']=='middle'],
              df3['science'][df3['ses']=='high'])
print('stat=%.3f,p=%.3f'%(stat,p))
if p>0.05:
    print('Probably the same distribution')
else:
    print('Probably different distribution')

stat=16.782,p=0.000
Probably different distribution
